File: src/main/java/gargoyle/netsquares/res/HybridCache.java
package gargoyle.netsquares.res;

import gargoyle.netsquares.util.Assert;

import java.lang.ref.Reference;
import java.lang.ref.SoftReference;
import java.util.HashMap;
import java.util.Map;

@SuppressWarnings("Duplicates")
public class HybridCache<K, V> {
    private final Map<K, Reference<V>> cache = new HashMap<>();
    private final Map<Class<? extends V>, Loader<K, ? extends V>> loaders = new HashMap<>();

    public HybridCache(Map<Class<? extends V>, Loader<K, ? extends V>> loaders) {
        this.loaders.putAll(loaders);
    }

    public <T extends V> T get(NSContext context, K name, Class<T> type, String suffix) {
        Loader<K, T> loader = getLoader(type);
        Assert.assertNotNull(loader, "loader is not known for {0}", type);
        return get(context, name, loader, suffix);
    }

    @SuppressWarnings("unchecked")
    private <T extends V> Loader<K, T> getLoader(Class<T> type) {
        Loader<K, ? extends V> Loader = loaders.get(type);
        return (Loader<K, T>) Loader;
    }

    @SuppressWarnings("unchecked")
    private <T extends V> T get(NSContext context, K name, Loader<K, T> loader, String suffix) {
        Assert.assertNotNull(name, "name is null");
        Assert.assertNotNull(loader, "loader is null");
        Reference<V> reference = cache.get(name);
        if (reference != null) {
            V value = reference.get();
            if (value != null) {
                return (T) value;
            }
        }
        try {
            T value = loader.load(context, name, suffix);
            Assert.assertNotNull(value, "loader did not load");
            cache.put(name, new SoftReference<>(value));
            return value;
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }
}

File: src/main/kotlin/gargoyle/netsquares/res/HybridCache.kt
package gargoyle.netsquares.res

import java.lang.ref.Reference
import java.lang.ref.SoftReference
import kotlin.reflect.KClass

class HybridCache<K : Any, V : Any>(loaders: Map<KClass<out V>, Loader<K, out V>>?) {
    private val cache: MutableMap<K, Reference<V?>> = mutableMapOf()
    private val loaders: MutableMap<KClass<out V>, Loader<K, out V>> = mutableMapOf()

    init {
        this.loaders.putAll(loaders!!)
    }

    operator fun <T : V> get(context: NSContext, name: K, type: KClass<T>, suffix: String): T {
        val loader = getLoader(type)
        return get(context, name, loader, suffix)
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T : V> getLoader(type: KClass<T>): Loader<K, T> {
        return loaders[type]!! as Loader<K, T>
    }

    @Suppress("UNCHECKED_CAST")
    private operator fun <T : V> get(context: NSContext, name: K, loader: Loader<K, T>, suffix: String): T {
        val reference = cache[name]
        if (reference != null) {
            val value = reference.get()
            if (value != null) {
                return value as T
            }
        }
        return try {
            val value = loader.load(context, name, suffix)
            cache[name] = SoftReference(value)
            value
        } catch (e: Exception) {
            throw RuntimeException(e.message, e)
        }
    }
}
